recirq/qcqmc/trial_wf_test.py

```diff
@@ -0,0 +1,387 @@
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    1.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                ]
+            ),
+        ),
+    ],
+)
+def test_qchem_pp_runs(
+    initial_two_body_qchem_amplitudes,
+    expected_ansatz_qubit_wf,
+    fixture_4_qubit_ham: HamiltonianData,
+):
+    params = PerfectPairingPlusTrialWavefunctionParams(
+        name="pp_test_wf_qchem",
+        hamiltonian_params=fixture_4_qubit_ham.params,
+        heuristic_layers=tuple(),
+        initial_orbital_rotation=None,
+        initial_two_body_qchem_amplitudes=initial_two_body_qchem_amplitudes,
+        do_optimization=False,
+    )
+
+    trial_wf = build_pp_plus_trial_wavefunction(
+        params,
+        dependencies={fixture_4_qubit_ham.params: fixture_4_qubit_ham},
+        do_print=False,
+    )
+
+    ansatz_qubit_wf = _get_ansatz_qubit_wf(
+        ansatz_circuit=trial_wf.ansatz_circuit,
+        ordered_qubits=params.qubits_jordan_wigner_ordered,
+    )
+
+    np.testing.assert_array_almost_equal(ansatz_qubit_wf, expected_ansatz_qubit_wf)
+
+
+def test_qchem_conversion_negative(fixture_4_qubit_ham: HamiltonianData):
+    qchem_amplitudes = np.asarray(-0.1)
+
+    two_body_params = get_two_body_params_from_qchem_amplitudes(qchem_amplitudes)
+
+    assert two_body_params.item() < 0
+
+    params = PerfectPairingPlusTrialWavefunctionParams(
+        name="pp_test_wf_qchem_neg",
+        hamiltonian_params=fixture_4_qubit_ham.params,
+        heuristic_layers=tuple(),
+        initial_orbital_rotation=None,
+        initial_two_body_qchem_amplitudes=qchem_amplitudes,
+        do_optimization=False,
+    )
+
+    trial_wf = build_pp_plus_trial_wavefunction(
+        params,
+        dependencies={fixture_4_qubit_ham.params: fixture_4_qubit_ham},
+        do_print=False,
+    )
+
+    ansatz_qubit_wf = _get_ansatz_qubit_wf(
+        ansatz_circuit=trial_wf.ansatz_circuit,
+        ordered_qubits=params.qubits_jordan_wigner_ordered,
+    )
+
+    assert any(ansatz_qubit_wf < 0)
+
+
+def gen_random_restricted_ham(n_orb: int) -> RestrictedHamiltonian:
+    """8-fold symmetry restricted hamiltonian"""
+    h1e = np.random.random((n_orb,) * 2)
+    h1e = h1e + h1e.T
+    h2e = np.random.random((n_orb,) * 4)
+    h2e = h2e + h2e.transpose(2, 3, 0, 1)
+    h2e = h2e + h2e.transpose(3, 2, 1, 0)
+    h2e = h2e + h2e.transpose(1, 0, 2, 3)
+    h2e = np.asarray(h2e.transpose(0, 2, 3, 1), order="C")
+    fqe_ham = RestrictedHamiltonian((h1e, np.einsum("ijlk", -0.5 * h2e)))
+    return fqe_ham
+
+
+def get_fd_grad(
+    n_orb,
+    n_elec,
+    one_body_params,
+    two_body_params,
+    ham,
+    initial_wf,
+    dtheta=1e-4,
+    restricted=False,
+):
```

Comment:
What is this?  Should this be in the test file?  Also, it's missing type info and docstring.

Reply:
It's the finite difference gradient of the ansatz wrt to the parameters. It's only used here for testing against the analytic gradients. I added the docs and types.